Allow to set a minimum particle per cell limit

gassmoeller opened the pull request:
This PR is more complicated than #791 and #792 and there remains a minor issue that needs to be addressed before the merge. This PR addresses a common problem in particle in cell codes if the number of particles in a particular cell decreases or even falls to zero (e.g. due to inflow boundary conditions, or because the mesh is refined several times in a formerly coarse region).
It proposes a relatively simple (certainly not optimal) solution in that the user can prescribe a minimum number of particles per cell. If the number of particles in a cell falls below this number a certain number of particles (the difference of actual particle count and set minimal particle count) is generated randomly inside this cell, and their properties are determined depending on the selected properties. For some properties it is more appropriate to initialize them to zero, for others it is more appropriate to interpolate the property from the surrounding particles.
The remaining issue concerns the uniqueness of the newly generated particle ids. In the current structure there is no simple way of asking "What is the next free and globally unique id for this new particle?". The new particles might be generated in parallel on separate processes. The only reasonable idea so far was to change the particle id from one unsigned integer that needs to be globally unique to a pair of integers, denoting the "generator" mpi_rank and locally unique particle_id (props to @jdannberg for this idea). If you agree, I would proceed with this in a separate PR first.

bangerth commented on include/aspect/particle/property/function.h:
```diff
@@ -69,6 +69,18 @@ namespace aspect
+           * Returns an enum, which determines how this particle property is
+           * initialized for particles that are created later than the initial
+           * particle generation, e.g. to balance the particle load or prevent
+           * empty cells. The current implementation returns
```
With "current implementation", do you mean the "implementation of this function in the current class"? I think the latter is clearer. Same in the parallel classes.

bangerth commented on include/aspect/particle/property/interface.h:
```diff
@@ -59,6 +61,31 @@ namespace aspect
+       * have been added later than the initial particle creation, e.g. to
+       * improve the load balance or to prevent empty cells.
+       */
+      enum InitializeLateParticles
```
This enum is awkwardly named because its first word is a verb -- but it  doesn't actually *do* anything. Maybe call it `InitializationModeForLateParticles`? This would also suggest renaming the `late_initialize_mode` function to `late_initialization_mode`.

bangerth commented on include/aspect/particle/world.h:
```diff
@@ -268,10 +269,19 @@ namespace aspect
+         * refinement, after MPI transfer of particles and after particle
+         * movement.
          */
         unsigned int max_particles_per_cell;
```
Out of curiosity, the documentation doesn't say what happens when the maximum is exceeded. Would it be worth saying what happens in that case here in the documentation?

bangerth commented on source/particle/property/interface.cc:
```diff
@@ -130,6 +140,49 @@ namespace aspect
+        for (typename std::list<std_cxx11::shared_ptr<Interface<dim> > >::const_iterator
+             p = property_list.begin(); p!=property_list.end(); ++p, ++property_index)
+          {
+            if ((*p)->late_initialize_mode() == InitializeLateParticles::initialize)
```
I think a better structure would replace the repeated `if - if else - if else` statements by a `switch -case -case` approach.

bangerth commented on source/particle/world.cc:
```diff
@@ -206,6 +210,88 @@ namespace aspect
+                  }
+
+                // Remove particles if necessary
+                if ((max_particles_per_cell > 0) && (particles_in_cell > max_particles_per_cell))
```
this case is mutually exclusive to the above case -- maybe insert an `else` before the `if`?

gassmoeller replied:
Not in the current form. But the logic is confusing, I agree. I reformulated that into an if-else if.

bangerth commented on source/particle/world.cc:
```diff
@@ -206,6 +210,88 @@ namespace aspect
+                    for (typename std::multimap<types::LevelInd, Particle<dim> >::iterator particle = particles_in_cell.first;
+                         particle != particles_in_cell.second; ++particle_index)
+                      {
+                        if (particle_index % GeometryInfo<dim>::max_children_per_cell != 0)
```
I don't think I understand this piece of code. You want to remove `particles_in_cell - max_particles_per_cell` particles here. `particle_index` runs through all particles on this cell. What is the connection between `particle_index` and the maximal number of children of a cell, and why does this condition guarantee the removal of the correct number of particles?

gassmoeller replied:
The procedure was different. When I wrote this code I assumed that the maximum number of particles would exceed the maximum number when coarsening happens, so to keep the particle density constant I would need to reduce the number of particles by a factor of 2^dim (the maximum number of children of a cell). Now with the new code it is awkward to have particle removal work differently from particle creation, so I will rewrite this to only remove `particles_in_cell - max_particles_per_cell` particles. However, I will need to remove random particles from the cell, because just removing from the beginning or the end of the particle list might look awkward for ordered particle distributions (suddenly a part of the cell will be cleared of particles, and another part still has the old particle density).

bangerth replied:
Ok. Just draw randomly which particles you want to remove in that case. I think you already have a random number generator lying around somewhere, then you just need to create an object with a uniform integer distribution between zero and the then current number of particles.

bangerth commented on source/postprocess/tracers.cc:
```diff
@@ -262,18 +262,19 @@ namespace aspect
+                             "Lower limit for particle number per cell. This limit is "
+                             "useful for adaptive meshes to prevent fine cells from being empty "
+                             "of particles. It will be checked and enforced after mesh "
+                             "refinement and after particle movement.");
```
Can you add what happens during enforcement (with possible links to what the various property initializers do)? Same below for the maximal number of particles.

bangerth commented on source/postprocess/tracers.cc:
```diff
@@ -325,7 +326,13 @@ namespace aspect
+
+          AssertThrow(min_tracers_per_cell <= max_tracers_per_cell,
+                      ExcMessage("Please select a 'Minimum tracers per cell' parameter "
+                                 "that is smaller or equal to the 'Maximum tracers per cell' parameter."));
```
smaller -> smaller THAN

bangerth commented:
The way this would typically be handled would be like this in `apply_particle_per_cell_bounds()`, assuming `global_next_particle_index` is already available:
```
  types::particle_index need_new_particles = 0;
  for (cell=...)
    if (...need new particles...)
      need_new_particles += ...;

  std::vector<types::particle_index> global_need_new_particles (n_mpi_processors);
  MPI_Allgather (&need_new_particles, 1, appropriate_datatype,
                           &global_need_new_particles[0], 1, appropriate_datatype,
                           mpi_communicator);

  types::particle_index my_first_new_particle_index = global_next_particle_index;
  for (unsigned int p=0; p<MPI::Utilities::this_mpi_process(...); ++p)
    my_first_new_particle_index += global_need_new_particles[p];
```

You can then start enumerating the new particles starting with `my_first_new_particle_index` without stepping on other processors.

This requires a first "scan" loop to determine who needs how many people, and one global exchange of these numbers, but I think the effort for this will be negligible.

bangerth commented on include/aspect/particle/generator/interface.h:
```diff
@@ -121,6 +138,15 @@ namespace aspect
+           * will be used to generate random particle locations.
+           */
+          boost::mt19937            random_number_generator;
+          boost::uniform_01<double> uniform_distribution_01;
```
By the way, these distribution objects have no state and are cheap to create. You may as well do that in the place where you use it.

gassmoeller replied:
I could, but I would like to have a deterministic behavior, to keep the tests working (currently many tests are failing because I moved this part to the interface but have another RNG in probability_density_function). That is why I initialize them once with a static seed in the constructor. Would that be possible if I create them in a member function?

bangerth replied:
Right, you need to keep the random number generator here. But you can move the distribution to wherever it is needed. When you draw a random number from the distribution, it just draws a number from the RNG and transforms it somehow. The sequence of numbers from the RNG remains deterministic.

gassmoeller commented:
I addressed all comments and also implemented your suggested solution and it seems to work good. Have not tested the scaling though, but I think you are right, it should be no bottleneck. Please have another look, or merge.
